cognite/client/_api/datapoints.py

```diff
@@ -1512,19 +1513,17 @@ def _validate_and_format_datapoints(
         datapoints: list[dict[str, Any]] | list[tuple[int | float | datetime, int | float | str]],
     ) -> list[tuple[int, Any]]:
         assert_type(datapoints, "datapoints", [list])
-        assert len(datapoints) > 0, "No datapoints provided"
+        if not datapoints:
+            raise ValueError("No datapoints provided")
         assert_type(datapoints[0], "datapoints element", [tuple, dict])
 
-        valid_datapoints = []
         if isinstance(datapoints[0], tuple):
-            valid_datapoints = [(timestamp_to_ms(t), v) for t, v in datapoints]
-        elif isinstance(datapoints[0], dict):
-            for dp in datapoints:
-                dp = cast(Dict[str, Any], dp)
-                assert "timestamp" in dp, "A datapoint is missing the 'timestamp' key"
-                assert "value" in dp, "A datapoint is missing the 'value' key"
-                valid_datapoints.append((timestamp_to_ms(dp["timestamp"]), dp["value"]))
-        return valid_datapoints
+            return [(timestamp_to_ms(t), v) for t, v in datapoints]
+        datapoints = cast(List[Dict[str, Any]], datapoints)
+        try:
+            return [(timestamp_to_ms(dp["timestamp"]), dp["value"]) for dp in datapoints]
+        except KeyError:
+            raise KeyError("A datapoint is missing one or both keys ['value', 'timestamp'].")
```

Comment:
I think the error message can be could be made more helpful by capturing all the failed datapoints, and display something like missing x of y timestamps and z of y values.

Reply:
I agree, but I don't think it belongs in this PR which seeks to remove wrong usage of `assert`s and `AssertionError`.